src/chip8/chip8.1
.TH chip8 6

.SH NAME
chip8 \- CHIP-8 emulator

.SH SYNOPSIS
.B chip8
[\fB\-h\fR | \fB\-\-help\fR]
[\fB\-v\fR | \fB\-\-version\fR]
[\fB\-\-hex\fR]
[\fB\-\-mute\fR]
.IR file ...

.SH DESCRIPTION
.B chip8
is another implementation of the CHIP-8 interpreted language and virtual
machine. It is designed as a modular application, is small, lightweight,
fast and multiplatform.

When the application is started, the ROM data contained in the file
.IR file
is loaded and then executed by the emulated virtual machine.

.SH OPTIONS
.TP
.B \-h ", " \-\-help
Shows the help message listing possible flags for the program.

.TP
.B \-v ", " \-\-version
Shows the installed version of the emulator.

.TP
.B \-\-hex
If provided, will consider the given file as an
.B hexadecimal file
instead of a binary file. An
.B hexadecimal file
is only composed of hexadecimal characters encoded in a human encoding format
such as US-ASCII or UTF-8, as an example the digits 0-9 and the letters A-F.
The emulator won't distinct between uppercase A-F and lowercase a-f although
the former is preferred. See below on section
.B ROMs
for more information on that.

.TP
.B \-\-mute
If provided, the emulator won't make any sound, which is useful for people
who don't want to play beeper sounds.

.SH ROMs
This emulator is compatible with CHIP-8 and SCHIP ROMs. A ROM is a file that
contains the opcodes that the virtual machine will run. There are two types of
ROMs:
.B binary ROMs
and
.B hexadecimal ROMs\fR.

By default ROM files are expected to be binary ROMs.
.B Binary ROMs
contains the program encoded as binary data. This means that every two bytes
in the ROM file will be mapped to a particular opcode executed by the virtual
machine (or to data used by some opcodes such as sprites). This is the default
behaviour since it produces the smallest possible files, the fastest to load
and the easiest to obtain (probably coming from a dump for an actual ROM chip).

On the other side,
.B hexadecimal ROMs
are actually
.B text files
encoded as ASCII or UTF-8 where every character represents an hexadecimal
character such as 0-9 or A-F. Note that because the file is encoded as ASCII
or UTF-8 and because the characters are only hexadecimal to the human, the
actual bytes stored in the file should always be in the range 0x30-0x39 (for
\fI0\fR\-\fI9\fR characters), 0x41-0x46 (for \fIA\fR\-\fIF\fR characters) or
0x61-0x66 (for \fIa\fR\-\fIf\fR characters). When the ROM is loaded, each two
ASCII characters found in the file are converted to a single byte by
translating the hexadecimal representation into actual bits. As an example, if
the sequence \fI6F\fR is read (0x36 0x46), the emulator translates it to the
byte 0x6F in the virtual memory.  Needless to say, this method requires more
time to start up the emulation since the ROM needs to be translated to binary,
although it will probably be the easiest to use for newcomers.

.SH BUGS
In case you find a bug, you can file it at our official issue tracker. The
canonical URL for our issue tracker is
<\fIhttp://github.com/danirod/chip8/issues\fR>.

.SH AUTHORS
This package is developed and maintained by
Dani Rodriguez (<\fIdanirod@outlook.com\fR>).

.SH COPYRIGHT
Copyright (C) 2015-2016 Dani Rodriguez


NEWS
CHIP-8 news - History of user visible changes
Copyright (C) 2015-2016 Dani Rodríguez <danirod@outlook.com>

This file is about relevant changes for CHIP-8 users. Developers: Check
ChangeLog for an overview on relevant changes about the source code.

Version 0.2.0 - 2016-06-18
* Added support for SUPER-CHIP opcodes.
* Fixes high CPU usage.

Version 0.1.3 - 2016-03-28
* Fixed: emulator won't crash if the PC has no sound device.
* Added: --mute flag was added, forcing emulator to not buzz.
* Added: support for DPKG distributions using a .deb file.

Version 0.1.2 - 2016-03-24
* Added sound support.
* Added more flags: --version and --usage.
* Fixed a few bugs on some opcodes.

Version 0.1.1 - 2015-03-12
* Emulator doesn't crash now on exit on Linux and MacOS X (issue #3)
* Emulator now uses the default system graphics backend. For example, might
  choose DirectX on Windows.

Version 0.1.0 - 2015-03-09
* First release with initial emulation capabilities.
* All the opcodes for a standard CHIP-8 machine have been implemented.
* Graphical capabilities for the CHIP-8 emulator are provided using SDL2.


src/chip8/chip8.c
/*
 * chip8 is a CHIP-8 emulator done in C
 * Copyright (C) 2015-2016 Dani Rodríguez <danirod@outlook.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <lib8/cpu.h>
#include "libsdl.h"
#include <config.h>

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

/* Flag set by '--hex' */
static int use_hexloader;

/* Flag set by '--mute' */
static int use_mute;

/* Flag used by '--debug' */
static int use_debug;

/* getopt parameter structure. */
static struct option long_options[] = {
    { "help", no_argument, 0, 'h' },
    { "version", no_argument, 0, 'v' },
    { "hex", no_argument, &use_hexloader, 1 },
    { "mute", no_argument, &use_mute, 1 },
    { "debug", no_argument, &use_debug, 1 },
    { 0, 0, 0, 0 }
};

/**
 * Print usage. In case you use bad arguments, this will be printed.
 * @param name how is the program named, usually argv[0].
 */
static void
usage(const char* name)
{
    /* How many characters has Usage: %s? */
    int pad = strnlen(name, 10) + 7; // 7 = "Usage: "

    printf("Usage: %s [-h | --help] [-v | --version]\n", name);
    printf("%*c [--hex] [--mute] <file>\n", pad, ' ');
}

static char
hex_to_bin(char hex)
{
    if (hex >= '0' && hex <= '9')
        return hex - '0';
    hex &= 0xDF;
    if (hex >= 'A' && hex <= 'F')
        return 10 + (hex - 'A');
    return -1;
}

/**
 * Load a hex file.
 *
 * @param file file path.
 * @param machine data structure to load the HEX into.
 */
static int
load_hex(const char* file, struct machine_t* machine)
{
    FILE* fp = fopen(file, "rb");
    if (fp == NULL) {
        fprintf(stderr, "Cannot open ROM file.\n");
        return 1;
    }

    // Use the fseek/ftell/fseek trick to retrieve file size.
    fseek(fp, 0, SEEK_END);
    int length = ftell(fp);
    fseek(fp, 0, SEEK_SET);

    // Create a temporal buffer where to store the data.
    char* hexfile = malloc(length);
    if (hexfile == NULL) {
        return 1;
    }
    fread(hexfile, length, 1, fp);
    fclose(fp);

    int mempos = 0x200;
    if (length & 0x01) length--;
    for (int i = 0; i < length; i += 2)
    {
        char hi = hex_to_bin(hexfile[i]);
        char lo = hex_to_bin(hexfile[i + 1]);
        if (hi == -1 || lo == -1) {
            free(hexfile);
            return 1;
        }

        machine->mem[mempos++] = hi << 4 | lo;
        if (mempos > 0xFFF)
            break;
    }

    free(hexfile);
    return 0;
}

/**
 * Load a ROM into a machine. This function will open a file and load its
 * contents into the memory from the provided machine data structure.
 * In compliance with the specification, ROM data will start at 0x200.
 *
 * @param file file path.
 * @param machine machine data structure to load the ROM into.
 */
static int
load_rom(const char* file, struct machine_t* machine)
{
    FILE* fp = fopen(file, "rb");
    if (fp == NULL) {
        fprintf(stderr, "Cannot open ROM file.\n");
        return 1;
    }

    // Use the fseek/ftell/fseek trick to retrieve file size.
    fseek(fp, 0, SEEK_END);
    int length = ftell(fp);
    fseek(fp, 0, SEEK_SET);

    // Check the length of the rom. Must be as much 3584 bytes long, which
    // is 4096 - 512. Since first 512 bytes of memory are reserved, program
    // code can only allocate up to 3584 bytes. Must check for bounds in
    // order to avoid buffer overflows.
    if (length > 3584) {
        fprintf(stderr, "ROM too large.\n");
        return 1;
    }

    // Everything is OK, read the ROM.
    fread(machine->mem + 0x200, length, 1, fp);
    fclose(fp);
    return 0;
}

static int
load_data(char* file, struct machine_t* mac)
{
    if (use_hexloader == 0) {
        return load_rom(file, mac);
    } else {
        return load_hex(file, mac);
    }
}

int
main(int argc, char** argv)
{
    struct machine_t mac;

    /* Parse parameters */
    int indexptr, c;
    while ((c = getopt_long(argc, argv, "hv", long_options, &indexptr)) != -1) {
        switch (c) {
            case 'h':
                usage(argv[0]);
                exit(0);
            case 'v':
                printf("%s\n", PACKAGE_STRING);
                exit(0);
                break;
            case 0:
                /* A long option is being processed, probably --hex. */
                break;
            default:
                exit(1);
        }
    }

    /*
     * optind should have the index of next parameter in argv. It should be
     * the name of the file to read. Therefore, should complain if file is not
     * given.
     */
    if (optind >= argc) {
        fprintf(stderr, "%1$s: no file given. '%1$s -h' for help.\n", argv[0]);
        exit(1);
    }

    /* Initialize SDL Context. */
    if (init_context()) {
        fprintf(stderr, "Error initializing SDL graphical context:\n");
        fprintf(stderr, "%s\n", SDL_GetError());
        return 1;
    }
    if (!try_enable_sound()) {
        fprintf(stderr, "Couldn't enable sound.\n");
        use_mute = 1;
    }

    /* Init emulator. */
    srand(time(NULL));
    if (use_debug) {
        set_debug_mode(1);
    }
    init_machine(&mac);
    mac.keydown = &is_key_down;
    if (!use_mute) {
        mac.speaker = &update_speaker;
    }
    load_data(argv[optind], &mac);


    int last_ticks = SDL_GetTicks();
    int last_delta = 0, step_delta = 0, render_delta = 0;
    while (!is_close_requested()) {
        /* Update timers. */
        last_delta = SDL_GetTicks() - last_ticks;
        last_ticks = SDL_GetTicks();
        step_delta += last_delta;
        render_delta += last_delta;

        /* Opcode execution: estimated 1000 opcodes/second. */
        while (step_delta >= 1) {
            step_machine(&mac);
            step_delta--;
        }

        /* Update timed subsystems. */
        update_time(&mac, last_delta);

        /* Render frame every 1/60th of second. */
        while (render_delta >= (1000 / 60)) {
            render_display(&mac);
            render_delta -= (1000 / 60);
        }

        /* Hack to reduce CPU usage :D
         * Maybe not best way but it works!! */
        SDL_Delay(1);
    }

    /* Dispose SDL context. */
    destroy_context();

    return 0;
}
